heartbeat/monitors/resolver.go

```diff
@@ -0,0 +1,102 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package monitors
+
+import (
+	"fmt"
+	"net"
+)
+
+// Resolver lets us define custom DNS resolvers similar to what the go stdlib provides, but
+// potentially with custom functionality
+type Resolver interface {
+	// ResolveIPAddr is an analog of net.ResolveIPAddr
+	ResolveIPAddr(network string, host string) (*net.IPAddr, error)
+	// LookupIP is an analog of net.LookupIP
+	LookupIP(host string) ([]net.IP, error)
+}
+
+// stdResolver uses the go std library to perform DNS resolution.
+type stdResolver struct{}
+
+// We only ever need one instance of this
+var StdResolver = stdResolver{}
+
+func (s stdResolver) ResolveIPAddr(network string, host string) (*net.IPAddr, error) {
+	return net.ResolveIPAddr(network, host)
+}
+
+func (s stdResolver) LookupIP(host string) ([]net.IP, error) {
+	return net.LookupIP(host)
+}
+
+type InvalidIPStringError struct {
+	ipString string
+}
+
+// staticResolver allows for a custom in-memory mapping of hosts to IPs, it ignores network names
+// and zones.
+type staticResolver struct {
+	mapping map[string][]net.IP
+}
+
+func CreateStaticResolver() staticResolver {
+	return staticResolver{mapping: map[string][]net.IP{}}
+}
+
+func (i InvalidIPStringError) Error() string {
+	return fmt.Sprintf("Could not parse IP from string %s", i.ipString)
+}
+
+func (s staticResolver) Add(hostname string, ip string) error {
```

Comment:
Might be a stupid question: is there a need for any TTL for IP addresses?

Reply:
No, an IP has no TTL, routing is determined dynamically by the next hop on your network, so there's nothing to expire.

Reply:
I was thinking rather about hostname and dynamically changed IP addresses..

Reply:
Sure, well, the goal here is for this to copy the interface of the go standard library and let us swap it out for test purposes, so adding functionality on top of that is probably not what we want here.

Reply:
Same here, looking forward to hear other voices.

Reply:
The DNS resolver in the go stdlib (and in Beats) do not cache results. Heartbeat has no custom DNS client yet. Caching and TTL is currently determined by the OS configuration.

I think I'm fine not having a TTL here on the method. If we want to introduce alternative implementations with caching functionality, then TTL should be given on init to the resolver.

Reply:
OK, given that we don't have any feature supporting TTLs, there doesn't seem to be a reason to add that in this PR